hnn_core/parallel_backends.py

```diff
@@ -0,0 +1,265 @@
+"""Parallel backends"""
+
+# Authors: Blake Caldwell <blake_caldwell@brown.edu>
+#          Mainak Jas <mainakjas@gmail.com>
+
+from warnings import warn
+
+_backend = None
+_loaded_dll = None
+
+
+class Joblib_backend(object):
+    """The Joblib_backend class.
+
+    Parameters
+    ----------
+    n_jobs : int
+        The number of jobs to start in parallel
+
+    Attributes
+    ----------
+    n_jobs : int
+        The number of jobs to start in parallel
+    """
+    def __init__(self, n_jobs=1):
+        self.type = 'joblib'
+
+        self.n_jobs = n_jobs
+        print("joblib will run over %d jobs" % (self.n_jobs))
+
+    def _parallel_func(self, func):
+        if self.n_jobs != 1:
+            try:
+                from joblib import Parallel, delayed
+            except ImportError:
+                warn('joblib not installed. Cannot run in parallel.')
+                self.n_jobs = 1
+        if self.n_jobs == 1:
+            my_func = func
+            parallel = list
+        else:
+            parallel = Parallel(self.n_jobs)
+            my_func = delayed(func)
+
+        return parallel, my_func
+
+    def __enter__(self):
+        global _backend
+        self._old_backend = _backend
+        _backend = self
+
+        return self
+
+    def __exit__(self, type, value, traceback):
+        global _backend
+        _backend = self._old_backend
+
+    def _clone_and_simulate(self, net, trial_idx):
+        # avoid relative lookups after being forked by joblib
+        from hnn_core.neuron import _neuron_network, _simulate_single_trial
+
+        if trial_idx != 0:
+            net.params['prng_*'] = trial_idx
+
+        neuron_net = _neuron_network(net.params)
+        dpl = _simulate_single_trial(neuron_net)
+
+        spikedata = neuron_net.get_data_from_neuron()
+
+        return dpl, spikedata
```

Comment:
We should address the standard API for returning data from simulations in a future PR. It will get longer than this (e.g. LFP) and it gets pickeld/unpickled with MPI_backend. Should it be a simdata dictionary for example?

Also, in my next PR that handles multiple trials, I plan to move this function outside Joblib_backend and import in mpi_child.py (they have the same code except).

Reply:
looking forward to further improvements in next PRs!